// @matterlabs/hardhat-zksync-verify/dist/src/zksync-block-explorer/verification-status-response.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerificationStatusResponse = void 0;
var VerificationStatusEnum;
(function (VerificationStatusEnum) {
    VerificationStatusEnum["SUCCESSFUL"] = "successful";
    VerificationStatusEnum["FAILED"] = "failed";
    VerificationStatusEnum["QUEUED"] = "queued";
    VerificationStatusEnum["IN_PROGRESS"] = "in_progress";
})(VerificationStatusEnum || (VerificationStatusEnum = {}));
class VerificationStatusResponse {
    constructor(response) {
        this.status = response.data.status;
        this.error = response.data.error;
        this.compilationErrors = response.data.compilationErrors;
    }
    errorExists() {
        return !!(this.error || this.compilationErrors);
    }
    getError() {
        let errors = '';
        if (this.error) {
            errors += this.error;
        }
        if (this.compilationErrors) {
            errors += this.compilationErrors.join('\n');
        }
        return errors;
    }
    isPending() {
        return this.status === VerificationStatusEnum.IN_PROGRESS;
    }
    isVerificationFailure() {
        return this.status === VerificationStatusEnum.FAILED;
    }
    isQueued() {
        return this.status === VerificationStatusEnum.QUEUED;
    }
    isVerificationSuccess() {
        return this.status === VerificationStatusEnum.SUCCESSFUL;
    }
}
exports.VerificationStatusResponse = VerificationStatusResponse;
