// @matterlabs/hardhat-zksync-upgradable/dist/src/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifiableContracts = exports.IMPL_CONTRACT_NOT_DEPLOYED_ERROR = exports.EVENT_NOT_FOUND_ERROR = exports.TOPIC_LOGS_NOT_FOUND_ERROR = exports.UPGRADE_VERIFY_ERROR = exports.PROXY_SOURCE_NAMES = exports.ZKVYPER_ARTIFACT_FORMAT_VERSION = exports.ZKSOLC_ARTIFACT_FORMAT_VERSION = exports.MANIFEST_DEFAULT_DIR = exports.FORMAT_TYPE_MINIMAL = exports.LOCAL_SETUP_ZKSYNC_NETWORK = exports.ERC1967_PROXY_JSON = exports.PROXY_ADMIN_JSON = exports.UPGRADABLE_BEACON_JSON = exports.BEACON_PROXY_JSON = exports.TUP_JSON = exports.ITUP_JSON = exports.PLUGIN_NAME = void 0;
exports.PLUGIN_NAME = '@matterlabs/hardhat-zksync-upgradable';
exports.ITUP_JSON = 'ITransparentUpgradeableProxy.json';
exports.TUP_JSON = 'TransparentUpgradeableProxy.json';
exports.BEACON_PROXY_JSON = 'BeaconProxy.json';
exports.UPGRADABLE_BEACON_JSON = 'UpgradeableBeacon.json';
exports.PROXY_ADMIN_JSON = 'ProxyAdmin.json';
exports.ERC1967_PROXY_JSON = 'ERC1967Proxy.json';
exports.LOCAL_SETUP_ZKSYNC_NETWORK = 'http://0.0.0.0:3050';
exports.FORMAT_TYPE_MINIMAL = 'minimal';
exports.MANIFEST_DEFAULT_DIR = '.upgradable';
exports.ZKSOLC_ARTIFACT_FORMAT_VERSION = 'hh-zksolc-artifact-1';
exports.ZKVYPER_ARTIFACT_FORMAT_VERSION = 'hh-zkvyper-artifact-1';
exports.PROXY_SOURCE_NAMES = [
    '@openzeppelin/contracts/proxy/transparent/TransparentUpgradeableProxy.sol',
    '@openzeppelin/contracts/proxy/transparent/ProxyAdmin.sol',
    '@openzeppelin/contracts/proxy/beacon/BeaconProxy.sol',
    '@openzeppelin/contracts/proxy/beacon/UpgradeableBeacon.sol',
    '@openzeppelin/contracts/proxy/ERC1967/ERC1967Proxy.sol',
];
exports.UPGRADE_VERIFY_ERROR = 'The verify plugin must be imported before the hardhat-upgrades plugin.' +
    'Import the plugins in the following order in hardhat.config.js:\n';
const TOPIC_LOGS_NOT_FOUND_ERROR = (topic, address) => `No logs found for event topic ${topic} at address ${address}\n` +
    `One of possible reasons can be that you are trying to verify a UUPS contract`;
exports.TOPIC_LOGS_NOT_FOUND_ERROR = TOPIC_LOGS_NOT_FOUND_ERROR;
const EVENT_NOT_FOUND_ERROR = (address, events) => `Could not find an event with any of the following topics in the logs for address ${address}: ${events.join(', ')}` +
    'If the proxy was recently deployed, the transaction may not be available on Block Explorer yet. Try running the verify task again after waiting a few blocks.';
exports.EVENT_NOT_FOUND_ERROR = EVENT_NOT_FOUND_ERROR;
exports.IMPL_CONTRACT_NOT_DEPLOYED_ERROR = 'The implementation contract was not previously deployed.\n' +
    'The useDeployedImplementation option was set to true but the implementation contract was not previously deployed on this network.';
exports.verifiableContracts = {
    erc1967proxy: { event: 'Upgraded(address)' },
    beaconProxy: { event: 'BeaconUpgraded(address)' },
    upgradeableBeacon: { event: 'OwnershipTransferred(address,address)' },
    transparentUpgradeableProxy: { event: 'AdminChanged(address,address)' },
    proxyAdmin: { event: 'OwnershipTransferred(address,address)' },
};
